File: March 2023/Day 10/Maximum Triplet product/maximum-triplet-product.java
//{ Driver Code Starts
//Initial Template for Java

/*package whatever //do not write package name here */

import java.io.*;
import java.util.*;


// } Driver Code Ends
//User function Template for Java

class Solution {
    Long maxTripletProduct(Long arr[], int n)
    {
        // Complete the function
        long mx1=Integer.MIN_VALUE;
        long mx2=mx1,mx3=mx1;
        long mn1=Integer.MAX_VALUE;
        long mn2=mn1;
        for(long val : arr) {
            if(mx1<=val) {
                mx3=mx2;
                mx2=mx1;
                mx1=val;
            } else if(mx2<=val) {
                mx3=mx2;
                mx2=val;
            } else if(mx3<=val) {
                mx3=val;
            }
            
            if(mn1>=val) {
                mn2=mn1;
                mn1=val;
            } else if(mn2>=val) {
                mn2=val;
            }
        }
        // System.out.println(mx1+" "+mx2+" "+mx3+" "+mn1+" "+mn2);
        return Math.max(mn1*mn2*mx1,mx1*mx2*mx3);
    }
}

//{ Driver Code Starts.

// Driver class
class Array {
    
    // Driver code
	public static void main (String[] args) throws IOException{
		// Taking input using buffered reader
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		int testcases = Integer.parseInt(br.readLine());
		
		// looping through all testcases
		while(testcases-- > 0){
		    
		    int sizeOfArray = Integer.parseInt(br.readLine());
		    
		    Long[] arr = new Long[sizeOfArray];
		    
		    String line = br.readLine();
		    String[] elements = line.trim().split("\\s+");
		    
		    for(int i = 0;i<sizeOfArray;i++){
		        arr[i] = Long.parseLong(elements[i]);
		    }
		    
		    Solution obj = new Solution();
		    Long res = obj.maxTripletProduct(arr, sizeOfArray);
		    System.out.println(res);
		}
	}
}


// } Driver Code Ends

File: March 2023/Day 10/Maximum Triplet product/maximum-triplet-product.cpp
//{ Driver Code Starts
//Initial Template for CPP

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
//User function Template for C++

class Solution {
  public:
    // Time Complexity :- O(n);
    // Space Complexity :- O(1);
    long long maxTripletProduct(long long arr[], int n)
    {
    	// Complete the function
    	long long mx1=INT_MIN;
        long long mx2=mx1,mx3=mx1;
        long long mn1=INT_MAX;
        long long mn2=mn1;
        for(int i=0;i<n;i++) {
            long long val = arr[i];
            if(mx1<=val) {
                mx3=mx2;
                mx2=mx1;
                mx1=val;
            } else if(mx2<=val) {
                mx3=mx2;
                mx2=val;
            } else if(mx3<=val) {
                mx3=val;
            }
            
            if(mn1>=val) {
                mn2=mn1;
                mn1=val;
            } else if(mn2>=val) {
                mn2=val;
            }
        }
        return max(mn1*mn2*mx1,mx1*mx2*mx3);
    }
};

//{ Driver Code Starts.

int main()
{
	int t;
	cin>>t;
	while(t--)
	{
	    int n,i;
	    cin>>n; long long arr[n];
	    for(i=0;i<n;i++)
	    cin>>arr[i];
	    Solution ob;
	    cout <<ob.maxTripletProduct(arr, n);
	    cout<<"\n";
	}
    return 0;
}
// } Driver Code Ends
